### PR TITLE
Update PE RNA-seq workflow

### PR DESCRIPTION
**Pull request by pavanvidem**
- Planned a single-end and a paired-end workflow for standard RNA-seq analysis
- Workflows use Cutadapt, STAR, featureCounts and DESeq2 follows the reference-based RNA-seq GTN tutorial

### CODE REVIEW

**Comment by lldelisle** [comment]
Hi,
It seems that this workflow follows the reference-based RNA-seq GTN but in a production workflow you don't want to do multiQC on each step but more only once at the end.
Also I feel like it is quite duplicated with the other 2 workflows.
What about:
- improving the current workflows by adding an optional subworkflow which would do the additional QCs that the GTN proposes
- propose to compute counts with featureCount in addition to the STAR counts.
- propose a DESeq2 step but this for me is tricky because you need to impose a format for your input like:
sample_treatment_batcheffect but this is fine.
I can help you with that if you want.

**Comment by pavanvidem** [comment]
Hey @lldelisle thanks for the feedback!
They are not exactly the same. I am trying to create a workflow that is closer to the tutorial. This workflow also includes DESeq2. I would also further extend it to plotting and filtering DE lists and so on. Maybe all the post-processing from DE genes list should become another workflow. We can discuss this later.

If we should merge both workflows, then I have the following questions/requests/suggestions. Maybe we can come to an agreement ;)
- Sure, I will merge all MultiQCs into one as you suggested
- The adapters in your workflow are required. Is it fine to make them optional?
- Why do we need Cufflinks FPKM step at the end? If we will include DESeq2 step, then we already direct users to DE analysis. Then can we remove the FPKM generation?
- Is featureCounts an additional output or alternative counts file? I think one of the count files should be enough.  I know that you did a comparison between those two approaches (STAR vs featureCounts) and if I remember correctly, the count estimated was similar. I always encourage featureCounts just because the counting step is independent of mapping. It is good to show in the training but maybe not required in a workflow. Which approach do you prefer?
- For DESeq2 step @wm75 prposed to use a file that maps from (whatever) filenames to treatments and batches. We will still use the [group-tags](https://training.galaxyproject.org/training-material/topics/galaxy-interface/tutorials/group-tags/tutorial.html) approach but they are provided by the user as an input file instead us extracting it from the file names. In this case, there is no restriction on how the file names should be.

**Comment by pavanvidem** [comment]
Good. I will merge the features from these workflows in to the existing.

**Comment by lldelisle** [comment]
Let me know if you need help

**Comment by pavanvidem** [comment]
what does it mean? https://github.com/galaxyproject/iwc/actions/runs/11258546316/job/31305455159?pr=544#step:7:1687
I also got in [another PR](https://github.com/galaxyproject/iwc/actions/runs/11256112453/job/31297295774?pr=556#step:7:2400).

I did not change anything major in my last commit.

**Comment by bgruening** [comment]
I think this is the error:

```

 Traceback (most recent call last):
  File "/tmp/tmpwsjhuvee/galaxy-dev/lib/galaxy/workflow/run.py", line 467, in replacement_for_connection
    replacement = step_outputs[output_name]
                  ~~~~~~~~~~~~^^^^^^^^^^^^^
KeyError: 'output'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/tmpwsjhuvee/galaxy-dev/lib/galaxy/workflow/run.py", line 238, in invoke
    incomplete_or_none = self._invoke_step(workflow_invocation_step)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/tmpwsjhuvee/galaxy-dev/lib/galaxy/workflow/run.py", line 322, in _invoke_step
    incomplete_or_none = invocation_step.workflow_step.module.execute(
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/tmpwsjhuvee/galaxy-dev/lib/galaxy/workflow/modules.py", line 2144, in execute
    collection_info = self.compute_collection_info(progress, step, all_inputs)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/tmpwsjhuvee/galaxy-dev/lib/galaxy/workflow/modules.py", line 517, in compute_collection_info
    collections_to_match = self._find_collections_to_match(progress, step, all_inputs)
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/tmpwsjhuvee/galaxy-dev/lib/galaxy/workflow/modules.py", line 537, in _find_collections_to_match
    data = progress.replacement_for_input(self.trans, step, input_dict)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/tmpwsjhuvee/galaxy-dev/lib/galaxy/workflow/run.py", line 442, in replacement_for_input
    replacement = self.replacement_for_connection(connection[0], is_data=is_data)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/tmpwsjhuvee/galaxy-dev/lib/galaxy/workflow/run.py", line 469, in replacement_for_connection
    raise modules.FailWorkflowEvaluation(
galaxy.workflow.modules.FailWorkflowEvaluation
```
So your workflow is corrupted

**Comment by pavanvidem** [comment]
@mblue9 can you please check what is going on here? The test summary says the following.

> The hosted runner encountered an error while running your job. (Error Type: Disconnect).

Btw, the tests are passing with planemo on my laptop against .org.

**Comment by mvdbeek** [comment]
Have you pinged the wrong person ?

**Comment by pavanvidem** [comment]
no :) Just checking whether you've seen such case before. 

I tried everything I could but no luck.

**Comment by mvdbeek** [comment]
So you really wanted to ping Maria ?

**Comment by mvdbeek** [comment]
Assuming you meant to ping me, you can rerun jobs, either in the interface or by closing and reopening the PR

**Comment by pavanvidem** [comment]
No. Now I understand. :facepalm:

**Comment by bgruening** [comment]
I have also restarted the test, no avail, something else needs to go on here.

**Comment by bgruening** [comment]
3rd time it at least turned red and crashed. Nothing suspicious to see for me:

> galaxy.workflow.run DEBUG 2024-10-17 06:53:41,017 [pN:main.1,p:5434,tN:WorkflowRequestMonitor.monitor_thread] Marking step 45 outputs of invocation 1 delayed (dependent step [51] delayed, so this step must be delayed)
galaxy.workflow.scheduling_manager DEBUG 2024-10-17 06:53:41,034 [pN:main.1,p:5434,tN:WorkflowRequestMonitor.monitor_thread] Workflow invocation [1] scheduled
galaxy.workflow.scheduling_manager DEBUG 2024-10-17 06:53:41,035 [pN:main.1,p:5434,tN:WorkflowRequestMonitor.monitor_thread] Attempting to schedule workflow invocation [2]
galaxy.workflow.run DEBUG 2024-10-17 06:53:41,127 [pN:main.1,p:5434,tN:WorkflowRequestMonitor.monitor_thread] Marking step 5 outputs of invocation 2 delayed (dependent step [4] delayed, so this step must be delayed)
galaxy.workflow.scheduling_manager DEBUG 2024-10-17 06:53:41,137 [pN:main.1,p:5434,tN:WorkflowRequestMonitor.monitor_thread] Workflow invocation [2] scheduled
galaxy.workflow.scheduling_manager DEBUG 2024-10-17 06:53:42,155 [pN:main.1,p:5434,tN:WorkflowRequestMonitor.monitor_thread] Attempting to schedule workflow invocation [1]
Error: Process completed with exit code 143.

**Comment by pavanvidem** [comment]
From the logs, the last running jobs are from STAR. The following featureCounts jobs did not get scheduled. Maybe STAR was taking too long to schedule. The file sizes are not so big (~50MB) for STAR. I subsampled the data to one of the smaller chromosomes which resulted in FASTQ files less than 1MB. I'm testing this locally, once I have those tests ready, I will push that data. I guess the STAR resource requirements mainly depends on the reference genome instead of FASTQ files. It should not make much difference. Let's see.

**Comment by lldelisle** [comment]
Maybe you can try to create another branch from main where you simply change the input fastqs and see if the test passes.

**Comment by pavanvidem** [comment]
> Maybe you can try to create another branch from main where you simply change the input fastqs and see if the test passes.

That seems to work: https://github.com/galaxyproject/iwc/pull/574. The tests are not passing because I did not update the tests but at least it ran the whole workflow.

**Comment by lldelisle** [comment]
I don't have time today but I will try to see what goes wrong in the workflow it can be an unconnected when or something like this. You can maybe remove parts and try CI until you find the issue.

**Comment by mvdbeek** [comment]
You mentioned you ran the workflow against .org, do you have a link ? If the workflow doesn't finish it's a resource issue and I might be able to tell ylou what consumed most memory

**Comment by pavanvidem** [comment]
here it is: https://usegalaxy.org/u/videmp/h/chr4-rnaseqpe

**Comment by pavanvidem** [comment]
Found some decent test data from yeast. Hopefully, any STAR jobs should not be an issue. 
The only issue is to indicate the runtime parameters of DESeq2 in tests. I am not sure if that is possible or not.

@lldelisle if it is not possible to write runtime parameters in tests, then the only way to have deseq2 in the workflow is with a fixed number of factors. We can go for a simple case with one factor with 2 factor levels. Then the input to the workflow is 2 separate collections, one for each factor level. 
Do you think is it a good idea to have such rigid DE analysis in this workflow so that we will have a simple but complete RNA-seq workflow? Or should we keep this as quantification workflow and create a second one for DE analysis, filtering and some plotting?

**Comment by pavanvidem** [comment]
In case you prefer a separate downstream analysis workflow, I created one here: https://github.com/galaxyproject/iwc/pull/582

**Comment by lldelisle** [comment]
I feel like it would be easier to indeed get 2 workflows.

**Comment by pavanvidem** [comment]
Thanks for the feedback! Then I will remove the DE part from this workflow and update the README. Then these two PRs will be ready to review.

**Comment by pavanvidem** [comment]
I am planning to create another two workflows (PE, SR) that connect the quantification workflows and this one. Those will be simple but complete workflows for 2 condition case. Do you think is it a good idea?

**Comment by lldelisle** [comment]
Maybe we can drop the number of samples as we don't have DE.

**Comment by lldelisle** [comment]
Hi @pavanvidem ,
I made a PR with my proposed changes: https://github.com/pavanvidem/iwc/pull/1

File: workflows/transcriptomics/rnaseq-pe/rnaseq-pe-tests.yml [hunk 1]
```diff
@@ -1,99 +1,86 @@
...
+    GTF with regions to exclude from FPKM normalization with Cufflinks: null
+    Compute StringTie FPKM: true
   outputs:
-    output_log:
+    MultiQC stats:
+      asserts:
+        has_text_matching:
+            expression: "SRR5085167\t0.10[0-9]*\t16.16[0-9]*\t43.61[0-9]*\t0.33[0-9]*\t0.31[0-9]*\t93.43\t0.11[0-9]*\t34.13\t0.19[0-9]*\t44.15[0-9]*\t\t\t\t\t\t"
+            expression: "SRR5085167_forward\t\t\t\t\t\t\t\t\t\t\t36.33[0-9]*\t46.0\t75.0\t75\t27.27[0-9]*\t0.39[0-9]*"
+            expression: "SRR5085167_reverse\t\t\t\t\t\t\t\t\t\t\t35.31[0-9]*\t46.0\t75.0\t75\t45.45[0-9]*\t0.39[0-9]*"
```
Comment (lldelisle) [inline workflows/transcriptomics/rnaseq-pe/rnaseq-pe-tests.yml:36]:
```suggestion
            expression: "SRR5085167_forward\t*36.33[0-9]*\t46.0\t75.0\t75\t27.27[0-9]*\t0.39[0-9]*"
            expression: "SRR5085167_reverse\t*35.31[0-9]*\t46.0\t75.0\t75\t45.45[0-9]*\t0.39[0-9]*"
```

File: workflows/transcriptomics/rnaseq-pe/rnaseq-pe-tests.yml [hunk 2]
```diff
@@ -1,99 +1,86 @@
...
-        GSM461177:
+        SRR5085167:
           asserts:
-            - that: "has_text"
-              text: "N_ambiguous	25107	5900	5518"
-            - that: "has_text"
-              text: "FBgn0010247	13	5	8"
-    HTS count like output:
+            has_text_matching:
+              expression: "YAL038W\tCDC19\tchrI\t\\+\t71786\t73288\t91.35[0-9]*\t3187.55[0-9]*\t3870.16[0-9]*"
```
Comment (lldelisle) [inline workflows/transcriptomics/rnaseq-pe/rnaseq-pe-tests.yml:54]:
```suggestion
              expression: "YAL038W\tCDC19\tchrI\t\\+\t71786\t73288\t92.5[0-9]*\t3273.9[0-9]*\t3900.[0-9]*"
```

File: workflows/transcriptomics/rnaseq-pe/rnaseq-pe-tests.yml [hunk 3]
```diff
@@ -1,99 +1,86 @@
...
       element_tests:
-        GSM461177_reverse:
+        SRR5085167:
           asserts:
-            has_size:
-              value: 5920766
-              delta: 600000
-        GSM461177_forward:
+            has_line:
+              line: "YAL038W	-	-	YAL038W	CDC19	-	chrI:71785-73288	-	-	3376.27	3155.33	3597.21	OK"
```
Comment (lldelisle) [inline workflows/transcriptomics/rnaseq-pe/rnaseq-pe-tests.yml:74]:
```suggestion
              line: "YAL038W	-	-	YAL038W	CDC19	-	chrI:71785-73288	-	-	3437.1	3211.44	3662.76	OK"
```

File: workflows/transcriptomics/rnaseq-pe/rnaseq-pe-tests.yml [hunk 4]
```diff
@@ -1,99 +1,86 @@
...
+    Transcripts Expression from Cufflinks:
+      element_tests:
+        SRR5085167:
           asserts:
-            has_size:
-              value: 5920766
-              delta: 600000
-    genes_expression_stringtie:
+            has_line:
+              line: "YAL038W_mRNA	-	-	YAL038W	CDC19	-	chrI:71785-73288	1503	102.307	3376.27	3155.33	3597.21	OK"
```
Comment (lldelisle) [inline workflows/transcriptomics/rnaseq-pe/rnaseq-pe-tests.yml:80]:
```suggestion
              line: "YAL038W_mRNA	-	-	YAL038W	CDC19	-	chrI:71785-73288	1503	102.837	3437.1	3211.44	3662.76	OK"
```

File: workflows/transcriptomics/rnaseq-pe/rnaseq-pe-tests.yml [hunk 5]
```diff
@@ -1,99 +1,86 @@
...
+    Counts Table:
       element_tests:
-        GSM461177:
+        SRR5085167:
           asserts:
-            has_text:
-              text: "FBgn0031217\tCG11377\tchr2L\t+\t102380\t104142\t1.963814\t33.024075\t62.009155"
+            has_line:
+              line: "YAL038W	1609"
```
Comment (lldelisle) [inline workflows/transcriptomics/rnaseq-pe/rnaseq-pe-tests.yml:86]:
```suggestion
              line: "YAL038W	1591"
```

**Review by lldelisle — COMMENTED** [review]
Change the tests results

File: workflows/transcriptomics/rnaseq-pe/rnaseq-pe-tests.yml [hunk 6]
```diff
@@ -1,99 +1,86 @@
...
+    GTF with regions to exclude from FPKM normalization with Cufflinks: null
+    Compute StringTie FPKM: true
   outputs:
-    output_log:
+    MultiQC stats:
+      asserts:
+        has_text_matching:
+            expression: "SRR5085167\t0.10[0-9]*\t16.16[0-9]*\t43.61[0-9]*\t0.33[0-9]*\t0.31[0-9]*\t93.43\t0.11[0-9]*\t34.13\t0.19[0-9]*\t44.15[0-9]*\t\t\t\t\t\t"
+            expression: "SRR5085167_forward\t*36.33[0-9]*\t46.0\t75.0\t75\t27.27[0-9]*\t0.39[0-9]*"
+            expression: "SRR5085167_reverse\t*35.31[0-9]*\t46.0\t75.0\t75\t45.45[0-9]*\t0.39[0-9]*"
```
Comment (lldelisle) [inline workflows/transcriptomics/rnaseq-pe/rnaseq-pe-tests.yml:36]:
```suggestion
        - that: "has_text_matching"
            expression: "SRR5085167\t0.10[0-9]*\t16.16[0-9]*\t43.61[0-9]*\t0.33[0-9]*\t0.31[0-9]*\t93.43\t0.11[0-9]*\t34.13\t0.19[0-9]*\t44.15[0-9]*\t\t\t\t\t\t"
        - that: "has_text_matching"
            expression: "SRR5085167_forward\t*36.33[0-9]*\t46.0\t75.0\t75\t27.27[0-9]*\t0.39[0-9]*"
        - that: "has_text_matching"
            expression: "SRR5085167_reverse\t*35.31[0-9]*\t46.0\t75.0\t75\t45.45[0-9]*\t0.39[0-9]*"
```

File: workflows/transcriptomics/rnaseq-pe/rnaseq-pe-tests.yml [hunk 7]
```diff
@@ -1,99 +1,88 @@
...
   outputs:
-    output_log:
+    MultiQC stats:
+      asserts:
+        - that: "has_text_matching"
+            expression: "SRR5085167\t0.10[0-9]*\t16.16[0-9]*\t43.61[0-9]*\t0.33[0-9]*\t0.31[0-9]*\t93.43\t0.11[0-9]*\t34.13\t0.19[0-9]*\t44.15[0-9]*\t\t\t\t\t\t"
+        - that: "has_text_matching"
+            expression: "SRR5085167_forward\t*36.33[0-9]*\t46.0\t75.0\t75\t27.27[0-9]*\t0.39[0-9]*"
+        - that: "has_text_matching"
+            expression: "SRR5085167_reverse\t*35.31[0-9]*\t46.0\t75.0\t75\t45.45[0-9]*\t0.39[0-9]*"
```
Comment (lldelisle) [inline workflows/transcriptomics/rnaseq-pe/rnaseq-pe-tests.yml:38]:
```suggestion
        - that: "has_text_matching"
          expression: "SRR5085167\t0.10[0-9]*\t16.16[0-9]*\t43.61[0-9]*\t0.33[0-9]*\t0.31[0-9]*\t93.43\t0.11[0-9]*\t34.13\t0.19[0-9]*\t44.15[0-9]*\t\t\t\t\t\t"
        - that: "has_text_matching"
          expression: "SRR5085167_forward\t*36.33[0-9]*\t46.0\t75.0\t75\t27.27[0-9]*\t0.39[0-9]*"
        - that: "has_text_matching"
          expression: "SRR5085167_reverse\t*35.31[0-9]*\t46.0\t75.0\t75\t45.45[0-9]*\t0.39[0-9]*"
```

File: workflows/transcriptomics/rnaseq-pe/rnaseq-pe-tests.yml [hunk 8]
```diff
@@ -1,99 +1,88 @@
...
+    Strandedness: stranded - forward
+    Compute Cufflinks FPKM: true
+    GTF with regions to exclude from FPKM normalization with Cufflinks: null
+    Compute StringTie FPKM: true
   outputs:
-    output_log:
+    MultiQC stats:
+      asserts:
+        - that: "has_text_matching"
+          expression: "SRR5085167\t0.10[0-9]*\t16.16[0-9]*\t43.61[0-9]*\t0.33[0-9]*\t0.31[0-9]*\t93.43\t0.11[0-9]*\t34.13\t0.19[0-9]*\t44.15[0-9]*\t\t\t\t\t\t"
```
Comment (lldelisle) [inline workflows/transcriptomics/rnaseq-pe/rnaseq-pe-tests.yml:34]:
```suggestion
          expression: "SRR5085167\t0.10[0-9]*\t16.45[0-9]*\t43.38[0-9]*\t0.32[0-9]*\t0.30[0-9]*\t93.75\t0.11[0-9]*\t34.29\t0.19[0-9]*\t17.6[0-9]*\t91.8[0-9]*\t30.0[0-9]*\t0.64[0-9]*\t43.5[0-9]*\t81.0[0-9]*\t72.6[0-9]*\t"
```

**Review by lldelisle — APPROVED** [review]
I am super happy of the result. This is a  good collaboration.

**Comment by lldelisle** [comment]
Tell me if you want to edit something before I merge it.

**Comment by pavanvidem** [comment]
All good from my side. Thanks for helping out!

**Comment by lldelisle** [comment]
I will make you a PR for the SR.

**Pull request merged**